galasa-managers-parent/galasa-managers-testingtools-parent/dev.galasa.selenium.manager/src/main/java/dev/galasa/selenium/internal/SeleniumEnvironment.java

```diff
@@ -0,0 +1,179 @@
+
+/**
+ * Selenium Environment is a generic environment for both local and remote drivers
+ * 
+ * 
+ * @author jamesdavies
+ *
+ */
+public class SeleniumEnvironment {
+	
+	private SeleniumManagerImpl seleniumManager;
+	private Path screenshotRasDirectory;
+		
+	private List<String> slots = new ArrayList<>();
+	private List<ISeleniumManager> drivers = new ArrayList<>();
+	
+	private IDynamicStatusStoreService dss;
+	private String runName;
+	
+	public SeleniumEnvironment(SeleniumManagerImpl seleleniumManager, Path screenshotRasDirectory) throws SeleniumManagerException {
+		this.seleniumManager = seleleniumManager;
+		this.screenshotRasDirectory = screenshotRasDirectory;
+		this.dss = seleleniumManager.getDss();
+		this.runName = seleniumManager.getFramework().getTestRunName();
+	}
+	
+	/**
+	 * Allocate the driver based on CPS configurations.
+	 * 
+	 * @param browser
+	 * @return
+	 * @throws SeleniumManagerException
+	 */
+	public ISeleniumManager allocateDriver(Browser browser) throws SeleniumManagerException {
+		ISeleniumManager driver;
+		Path driverRasDir = screenshotRasDirectory.resolve("driver_"+drivers.size());
+		// Get a slot or fail
+		String slotName = allocateSlot();
+		
+		try {
+			if (browser.equals(Browser.NOTSPECIFIED)) {
+				browser = Browser.valueOf(SeleniumDefaultDriver.get());
+			}
+			
+			switch(SeleniumWebDriverType.get()) {
+		    case ("local"):
+		    	driver = new LocalDriverImpl(browser, driverRasDir);
+		    	break;
+		    default:
+		    	driver =  new RemoteDriverImpl(this, seleniumManager, browser, slotName, driverRasDir);
+		    	break;
+			}
+		} catch (ConfigurationPropertyStoreException e) {
+			throw new SeleniumManagerException("Failed to fetch Driver type", e);
+		}
+		drivers.add(driver);
+		return driver;
+
+	}
+	
+	/**
+	 * A selenium specific slot to limit the number of selenium parallel runs.
+	 * 
+	 * @return
+	 * @throws SeleniumManagerException
+	 */
+	private String allocateSlot() throws SeleniumManagerException {
+		String slotKey = "driver.current.slots";
+		String slots = "";
+		String slotName = "";
+		int currentSlots = 0;
+		try {
+			slots = dss.get(slotKey);
+			if (slots != null) {
+				currentSlots = Integer.valueOf(slots);
+			}
+			
+			if (currentSlots >= SeleniumDriverMaxSlots.get()) {
+				throw new SeleniumManagerException("Failed to provsion. No slots avilable");
+			}			
+			
+			String slotNamePrefix = "SeleniumSlot_" + this.runName + "_";
+			
+			int counter = 0;
+			while ("".equals(slotName)) {
+				String slotNameAttempt = slotNamePrefix + counter;
+				if (dss.get("driver.slot." + slotNameAttempt)==null) {
+					// Found a slot name;
+					slotName = slotNameAttempt;
+					break;
+				}
+				
+				counter++;
+			}
+			
+		} catch (DynamicStatusStoreException | ConfigurationPropertyStoreException e) {
+			throw new SeleniumManagerException("Failed ot allocate slot", e);
+		}
+		
+		try {
+			dss.performActions(
```

Comment:
this should be merged with the above slot name change,   so they are all updated together or not at all

Reply:
Little confused on this one, as this is what that is doing. Looks up available slot names, then tries to update the number of slots at teh same time as incremementing the slot number

Reply:
problem is,  by the time you have checked the slot is not already in use,     and then reserve it,   another test may have already claimed it.   I know it is milliseconds,  but believe it will happen.

Instead of checking for it's existence, then reserve it,     you can do both at the same time,  by setting the old value to NULL.   that means,  the value does not already exist.